Extensive reorg/expansion of workspaces and runs sections (#1078)
hashicorp/terraform-website: content/source/docs/cloud/workspaces/vcs.html.md

---
layout: "cloud"
page_title: "VCS Connections - Workspaces - Terraform Cloud"
---

# Configuring Workspace VCS Connections

Any Terraform Cloud workspace can be connected to a version control system (VCS) repository that contains its Terraform configuration. You can assign a connection [when creating a workspace](./creating.html), and can modify that connection in the "Version Control" page of the workspace's settings.

For more information about how configuration versions and connected repositories work, see [Terraform Configurations in Terraform Cloud Workspaces](./configurations.html).

> -> **API:** See the [Update a Workspace endpoint](../api/workspaces.html#update-a-workspace) (`PATCH /organizations/:organization_name/workspaces/:name`).

Reviewer: Remove or move to the bottom of this section.

Author: (tabling)

To change the VCS settings for an existing workspace, choose "Version Control" from the workspace's "Settings" menu. The version control page includes the following settings:

- Which VCS repository to use.
- Which directories in the repository should trigger Terraform runs when they change.
- Which branch of the repository to use.
- Whether to download the repository's submodules.

For most of the settings on this page, you must save any changes with the "Update VCS settings" button at the bottom of the page.

> ## VCS Connection and Repository

Reviewer: Is the content in this section duplicated from somewhere else? It feels familiar to me. If so instead of duplicating link out if possible.

Author: UNFORTUNATELY, the product is kind of against me here. This interface is VERY similar to the flow on the "create a new workspace" page, but it has just enough critical differences that the descriptions can't be combined.

Author: (I tried. 😑)

The first item in the version control settings lets you select a new VCS repository or disconnect from the current repository.

> -> **Note:** Depending on the current status, this control can appear as a "Connect to version control" button or a "Change VCS connection" link.

Reviewer: Combine with the last sentence into one paragraph. 

```suggestion
Depending on the current status, this control can appear as a "Connect to version control" button or a "Change VCS connection" link.
```

> Connecting or disconnecting a VCS repository occurs on a separate page, split into three screens:

Reviewer: ```suggestion
Clicking this button or link will send you to a separate page split into three screens that will walk you through the process of connecting or disconnecting your VCS repository.
```

Author: ```suggestion
Clicking this control will navigate to a separate page, which will walk you through the repository configuration. This page is split into three screens.
```

1. On the first screen, choose your VCS provider (or choose "No VCS connection" to disconnect the workspace from version control).

>     -> **Note:** If you haven't added a VCS provider for your organization yet, choosing one here will prompt you to configure it. See [Connecting VCS Providers](../vcs/index.html) for more information. Only organization owners can configure VCS providers.

Reviewer: ```suggestion
    If you haven't added a VCS provider for your organization yet, choosing one here will prompt you to configure it. See [Connecting VCS Providers](../vcs/index.html) for more information. Only organization owners can configure VCS providers.
```

Author: Mmf. I put it in a box because the way it kicks you out of the flow to configure OAuth feels very dramatic in practice... but you're probably right.

1. On the second screen, choose a repository from the filterable list. This screen is skipped if you chose "No VCS connection".

    Some VCS providers limit the list's size. If a repository isn't listed, you can still choose it by name; scroll to the bottom of the list and enter its ID in the text field.

>     -> **Note:** For some VCS providers, this list includes a drop-down menu for changing which account's repositories are shown. Other providers combine all available accounts into a single list.

Reviewer: Combine with previous sentence to make a paragrph

```suggestion
    For some VCS providers, this list includes a drop-down menu for changing which account's repositories are shown. Other providers combine all available accounts into a single list.
```

Author: Well, agree about dropping the box, but IMO this should stay different paragraphs. Describing totally separate parts of the list's behavior.

1. On the third screen, confirm or cancel your choice. Confirmed changes will be saved immediately, and you will be returned to the "Version Control" settings page.

See also:

- [Creating Workspaces](./creating.html)
- [Connecting VCS Providers to Terraform Cloud](../vcs/index.html)

-> **API:** If you need to change VCS connections for many workspaces at once, consider automating the changes with the [Update a Workspace endpoint](../api/workspaces.html#update-a-workspace). This is most common when moving a VCS server, or when a vendor deprecates an older API version.

## Automatic Run Triggering

> -> **Note:** This setting only affects workspaces that specify a Terraform working directory. [The working directory setting](./settings.html#terraform-working-directory) can be found on the "General" settings page.

Reviewer: Maybe remove this note or combine it with paragraph 2? It feels a bit whiplash-y to think about workspaces that do, then workspaces that don't, and then workspaces that do again.

Author: Mm, yeah. Artifact of the previous layout of this info.

For workspaces that **don't** specify a Terraform working directory, Terraform Cloud assumes that the entire repository is relevant to the workspace. Any change will trigger a run.

> For workspaces that **do** specify a Terraform working directory, Terraform Cloud assumes that only _some_ content in the repository is relevant to the workspace. Only changes that affect the relevant content will trigger a run. By default, only the working directory is considered relevant.

Reviewer: ```suggestion
For workspaces that **do** specify a Terraform working directory, Terraform Cloud assumes that only some content in the repository is relevant to the workspace. Only changes that affect the relevant content will trigger a run. By default, only the working directory is considered relevant.
```

You can adjust this behavior in two ways:

- **Add more trigger directories.** Terraform Cloud will queue runs for changes in any of the specified trigger directories (including the working directory).

    For example, if you use a top-level `modules` directory to share Terraform code across multiple configurations, changes to the shared modules are relevant to every workspace that uses that repo. You can add `modules` as a trigger directory for each workspace to make sure they notice any changes to shared code.
- **Mark the entire repository as relevant.** If you set the "Automatic Run Triggering" setting to "Always Trigger Runs," Terraform Cloud will assume that anything in the repository might affect the workspace's configuration, and will queue runs for any change.

    This can be useful for repos that don't have multiple configurations but require a working directory for some other reason. It's usually not what you want for true monorepos, since it queues unnecessary runs and slows down your ability to provision infrastructure.

> -> **Note:** Trigger directories also apply to [speculative plans](./index.html#speculative-plans) on pull requests — Terraform Cloud won't queue plans for changes that aren't marked as relevant.

Reviewer: Should this go above the option to mark the whole repo as relevant?
